scipy/spatial/transform/_rotation_groups.py

```diff
@@ -0,0 +1,209 @@
+import numpy as np
+from . import Rotation
+from scipy.constants import golden as phi
+
+
+def build_group(generators):
+
+    # Perform a breadth-first search to find group elements of the binary
+    # quaternion group.
+    tol = 1E-10
+    identity = Rotation.from_quat([0, 0, 0, 1])
+    queue = [identity] + generators
+    visited = np.array([identity.as_quat()])
+    while queue:
+        node = queue.pop(0)
+        mindist = np.min(np.linalg.norm(node.as_quat() - visited, axis=1))
+        if mindist > tol:
+            visited = np.concatenate((visited, [node.as_quat()]))
+            for e in generators:
+                queue.append(node * e)
+
+    # Discard half the binary quaternion group to get the unique elements of
+    # the rotation group.
+    anchor = [1, 10, 100, 1000]
+    indices = np.where(np.dot(visited, anchor) > 0)[0]
+    visited = visited[indices]
+
+    # Clean up numerical noise for cleaner printing.
+    indices = np.where(np.abs(visited) < tol)
+    visited[indices] *= 0
+
+    return Rotation.from_quat(visited)
+
+
+def equispaced_semicircle(n):
+
+    thetas = np.linspace(0, np.pi, n + 1)[:n]
+    return np.sin(thetas), np.cos(thetas)
+
+
+class classproperty(object):
+
+    def __init__(self, fget):
+        self.fget = fget
+
+    def __get__(self, owner_self, owner_cls):
+        return self.fget(owner_cls)
+
+
+class RotationGroup:
```

Comment:
I have simply moved the `fundamental` method into `RotationGroup` for now.  A subclass is a nice idea though.  I will mark this PR as WIP.

Reply:
And actually it's super easy to do.

Reply:
After making it subclass of `Rotation`, can we keep only constructor of the form `__init__(self, group)`, where `group` is a string encoding the group. 

For example `RotationGroup("I")`, `RotationGroup("D4")`.

Reply:
If we make `RotationGroup` a subclass of `Rotation`, then type propagation will depend on the order of multiplications:

```
>>> g = RotationGroup("I")
>>> r = Rotation.from_quat([0, 0, 0, 1]) #identity rotation

>>> type(g * r)
<class 'RotationGroup'>
>>> type(r * g)
<class 'Rotation'>
```

Is this not confusing?

Reply:
I need to think about that. We can add another factor method `Rotation.create_group(...)`. But are we getting too many of these factory methods?

The `fundamental` method can be explained more or less clearly without the notion of the symmetry groups (am I correct?). I even think the name `reduce` might be appropriate. Probably `r.reduce(p)` should return `Rotation` containing `len(p)` rotations (`r` reduces `p`), but you might have other and better intuition about that.

Reply:
I like both the `Rotation.create_group(...)` and the `reduce` proposals.

Reply:
Yes, it looks like a decent design.

Do you have time to finish it in the coming days? I can work on it too, it's quite interesting.

Reply:
Yes, I can finish it soon.

I have made a `create_group` method which takes a string.  It ignores the `axis` parameter for the `I`, `O`, and `T` groups.  Is that OK?

I misread your comment and left it all in the `RotationGroup` class.  Do you want to get rid of this class and put everything in the `Rotation` class?

Reply:
> I misread your comment and left it all in the RotationGroup class. Do you want to get rid of this class and put everything in the Rotation class?

Yes, I think it's better this way. Now `RotationGroup` servers only as a namespace. But I suggest keeping the code within a separate file to avoid clogging `Rotation` farther.

> It ignores the axis parameter for the I, O, and T groups. Is that OK?

Up to you. But if it's not very difficult perhaps support it? I guess it can be achieved by premultiplication of the generated rotations by a constant rotation to change the axis.

I also suggest using `x, y, z, X, Y, Z` to select the axis, it better matches convention of `from_euler` and `as_euler`.

Reply:
> I also suggest using x, y, z, X, Y, Z to select the axis, it better matches 

I have changed the valid inputs to 'X', 'Y', and 'Z'.

The `I`, `O`, and `T` groups are already aligned to all the axes (x, y, and z).  I have added a note in the docstring that the axis parameter is ignored for these groups.

I have moved the `create_group` functionality into the `Rotation` class.  I don't know how to nicely split it across two files though.  The examples I can find on stackoverflow don't seem very elegant.

Other than that, I think all that we are missing is a good description for the `reduce` docstring.  You mentioned that it would be best to describe it without referring to rotation groups.  Do you have a suggestion?

Reply:
I have implemented your suggestion below.